src/unix/unix_c/unix_get_pw_gr_nam_id_job.c

```diff
@@ -84,6 +85,15 @@ static value alloc_group_entry(struct group *entry)
         job->buffer = (char *)lwt_unix_malloc(buffer_size);            \
         job->result = FUNC##_r(job->ARG, &job->entry, job->buffer,     \
                                buffer_size, &job->ptr);                \
+        while (job->result == -1 && errno == ERANGE)                   \
```

Comment:
I think this comparison with `-1` is incorrect. From the [man page](https://www.man7.org/linux/man-pages/man3/getpwnam.3.html):

> On success, `getpwnam_r()` and `getpwuid_r()` return zero, and set
> `*result` to `pwd`.  If no matching password record was found, these
> functions return `0` and store `NULL` in `*result`.  In case of error, an
> error number is returned, and `NULL` is stored in `*result`.

So it should probably be `job->result != 0`. Compare that also with `if (result)` down in `result##FUNC` later on in this file (although it would be clearer if it was `if (result != 0)`).

Reply:
As for `errno`, it's possible that `job->result` itself will be `ERANGE`. Otherwise, if still checking `errno`, there is this in the man page:

> If an error occurs, `errno` is set appropriately.  If one
> wants to check `errno` after the call, it should be set to zero before
> the call.

However, that's in the paragraph for the non-`_r` versions of these functions, so I wonder if the `_r` versions set `errno` at all, in which case `job->result` should be checked for `ERANGE`. I would check this locally in a small C program by calling these functions and triggering any failure, and examining the return value and `errno`.

Reply:
I have a new version (pushed) that should handle correctly:
- success case: result is `0` and buffer has not `NULL` (breaks and post-processing is handled by `result_##FUNC`)
- not found: result is `0` and buffer has `NULL` (same)
- error: result is non `0` and buffer has `NULL` (result is checked (rather than errno, as suffested by manpage) for ERGANE and either (a) it is post-processed by `result_##FUNC` of (b) buffer is freed, size is increased, buffer-size is checked, and we just retry)

It uses a lot more breaks than before. So I'm even less confident in the code.